index.js

```diff
@@ -125,6 +136,463 @@
+    return xs[xs.length - 1];
+  };
+
+  //  toPairs :: StrMap a -> [Pair String a]
+  var toPairs = function(obj) {
+    var pairs = [];
+    for (var k in obj) {
+      pairs.push([k, obj[k]]);
+    }
+    return pairs;
+  };
+
+  //  isFirstChild :: (Node, Node) -> Boolean
+  var isFirstChild = function(node, parent) {
+    return head(nonDelimiters(parent.children)) === node;
+  };
+
+  //  isRootFirstChild :: (Node, [Node]) -> Boolean
+  var isRootFirstChild = function(node, parents) {
+    return (
+      isRootChild(parents) &&
+      isFirstChild(node, head(parents))
+    );
+  };
+
+  //  isRoot :: [Node] -> Boolean
+  var isRoot = function(parents) {
+    return isEmpty(parents);
+  };
+
+  //  isRootChild :: [Node] -> Boolean
+  var isRootChild = function(parents) {
+    return parents.length === 1;
+  };
+
+  //  nonDelimiters :: [Node] -> [Node]
+  var nonDelimiters = function(nodes) {
+    return reject(nodes, isDelimiter);
+  };
+
+  //  arrayClone :: [a] -> [a]
+  var arrayClone = function(a) {
+    return a.slice(0);
+  };
+
+  //  isLeaf :: Node -> Boolean
+  var isLeaf = function(node) {
+    return isEmpty(node.children);
+  };
+
+  //  isDelimiter :: Node -> Boolean
+  var isDelimiter = function(t) {
+    return t.type === 'DELIMITER';
+  };
+
+  //  isContainer :: Node -> Boolean
+  var isContainer = function(t) {
+    return t.type === 'UNARY' || t.type === 'BINARY';
+  };
+
+  //  isVar :: Node -> Boolean
+  var isVar = function(t) {
+    return t.type === 'VARIABLE';
+  };
+
+  //  getConstraintText :: Node -> String
+  var getConstraintText = function(constraint) {
+    return stripNamespace(constraint.name);
+  };
+
+  //  traverse :: ((Node, [Node]) -> Void) -> Node -> Void
+  var traverse = function(f) {
+    return function(startNode) {
+      return (function recur(node, parents) {
+        f(node, parents);
+        var nodes = arrayClone(node.children);
+        for (var idx = 0; idx < nodes.length; idx += 1) {
+          recur(nodes[idx], parents.concat(node));
+        }
+        return node;
+      })(startNode, []);
+    };
+  };
+
+  //  traverseFind :: ((Node, [Node]) -> Nullable Node) -> Node
+  //                  -> Nullable Node
+  var traverseFind = function(f, startNode) {
+    return (function recur(node, parents) {
+      var r = f(node, parents);
+      if (r != null) {
+        return r;
+      }
+      var nodes = arrayClone(node.children);
+      for (var idx = 0; idx < nodes.length; idx += 1) {
+        r = recur(nodes[idx], parents.concat(node));
+        if (r != null) {
+          return r;
+        }
+      }
+      return null;
+    })(startNode, []);
+  };
+
+  //  traverseWith :: ((a, Node, Node) -> a) -> a -> Node -> Node -> a
+  var traverseWith = function(f, acc, node, parent) {
+    acc = f(acc, node, parent);
+    var nodes = arrayClone(node.children);
+    for (var idx = 0; idx < nodes.length; idx += 1) {
+      acc = traverseWith(f, acc, nodes[idx], node);
```

Comment:
I like to name recursive functions. My reasoning is that a function should depend on the outside world in as few ways as possible. The name of the identifier to which a function is bound is part of the world outside the function. The function's name, though, is part of the function's world.

I'm happy with either of the following names:

```javascript
var traverseWith = function traverseWith(f, acc, node, parent) { /* ... */ };
```

```javascript
var traverseWith = function recur(f, acc, node, parent) { /* ... */ };
```